### PR TITLE
Add flexible height mode to `BottomSheetController`

### PR DESCRIPTION
**Pull request by lcapkovic**
### Platforms Impacted
- [x] iOS
- [ ] macOS

### Description of changes

Currently our bottom sheet works in a "rigid" mode where its height is static and dragging it just translates the sheet view vertically on or off-screen.
Some clients need to pin views to the bottom of the sheet that are always visible (like a text field) so we need a flexible mode where dragging actually resizes the sheet, so bottom of the sheet content is always visible.

This PR adds the `isFlexibleHeight` property which enables the flexible height mode. The sheet will resize anywhere above collapsed state and stay rigid anywhere below it (to make sure we're not forcing the content to layout in small or zero heights.

### Verification
Default rigid mode:

https://user-images.githubusercontent.com/3610850/151105447-757a9359-69ae-47ba-9cff-f638f6f99851.mov

Flexible mode:

https://user-images.githubusercontent.com/3610850/151105461-199ebd9d-c123-49f5-b522-b3ad5dddaec4.mov

### Pull request checklist

This PR has considered:
- [x] Light and Dark appearances
- [x] iOS supported versions (all major versions greater than or equal current target deployment version)
- [x] VoiceOver and Keyboard Accessibility
- [x] Internationalization and Right to Left layouts
- [x] Different resolutions (1x, 2x, 3x)
- [x] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)
- [x] iPad [Pointer interaction](https://developer.apple.com/documentation/uikit/pointer_interactions)
- [ ] [SwiftUI](https://developer.apple.com/tutorials/swiftui) consumption (validation or new demo scenarios needed)
- [ ] Objective-C exposure (provide it only if needed)

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/fluentui-apple/pull/875)

### CODE REVIEW

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/BottomSheetDemoController.swift [hunk 1]
```diff
@@ -81,14 +86,37 @@ class BottomSheetDemoController: UIViewController {
...
         return personaListView
     }()
 
+    private lazy var expandedContentView: UIView = {
+        let view = UIView()
+        view.translatesAutoresizingMaskIntoConstraints = false
+        view.addSubview(personaListView)
+
+        let bottomView = UIView()
+        bottomView.backgroundColor = .systemTeal
```
Comment (rdeassis) [inline ios/FluentUI.Demo/FluentUI.Demo/Demos/BottomSheetDemoController.swift:103]: not sure I understand the reason for setting the backgroundColor to .systemTeal here...

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift [hunk 2]
```diff
@@ -97,6 +97,23 @@ public class BottomSheetController: UIViewController {
...
+    ///
+    /// When set to `true`, moving the sheet beyond the `.collapsed` state will resize it.
+    ///
+    /// Use flexible height if you have views attached to the bottom of `expandedContentView` that should always be visible.
+    @objc open var isFlexibleHeight: Bool = false {
+        didSet {
+            guard isViewLoaded else {
+                return
+            }
+            view.setNeedsLayout()
```
Comment (anandrajeswaran) [inline ios/FluentUI/Bottom Sheet/BottomSheetController.swift:113]: Why is this needed? I would have expected toggling this doesn't change anything about its current position and only affects what happens when interacted with? It sounds like this will also cause a one-time unanimated snap to allowed conditions that may be worth documenting?

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift [hunk 3]
```diff
@@ -741,20 +768,33 @@ public class BottomSheetController: UIViewController {
...
+        return max(collapsedSheetHeight, height)
+    }
+
+    // Height of the sheet in collapsed state
+    private var collapsedSheetHeight: CGFloat {
+        let visibleContentHeight = (collapsedContentHeight > 0) ? collapsedContentHeight : headerContentHeight
+        return currentResizingHandleHeight + visibleContentHeight + view.safeAreaInsets.bottom
+    }
+
+    private var currentResizingHandleHeight: CGFloat {
```
Comment (anandrajeswaran) [inline ios/FluentUI/Bottom Sheet/BottomSheetController.swift:796]: Minor: Do we have some convention on preferring computed vars vs. funcs? Or what the tradeoffs are?

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/BottomSheetDemoController.swift (hunk 1 above)
Comment (lcapkovic) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/BottomSheetDemoController.swift:103]: Yeah sorry, the demo app design should be a bit more intentional. I'll change it to something better.

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift (hunk 2 above)
Comment (lcapkovic) [reply ios/FluentUI/Bottom Sheet/BottomSheetController.swift:113]: The sheet frame needs to be recalculated, because everywhere except fully expanded state the frame will be different in flexible mode. For example if we're in collapsed state, setting this to true will resize the sheet so that its bottom is aligned with the bottom of the screen. Its offset wouldn't change in this case.

This won't cause a snap if the sheet is being interacted with because it'll be in the `.transitioning` state. In that case the interaction will automatically switch to calculating the frame using the new flexible / rigid mode.

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift (hunk 3 above)
Comment (lcapkovic) [reply ios/FluentUI/Bottom Sheet/BottomSheetController.swift:796]: Not that I'm aware of. I'd be very surprised if there were actual perf differences. Do you prefer a func here?

File: ios/FluentUI.Demo/FluentUI.Demo/Demos/BottomSheetDemoController.swift (hunk 1 above)
Comment (lcapkovic) [reply ios/FluentUI.Demo/FluentUI.Demo/Demos/BottomSheetDemoController.swift:103]: I updated it, also added new screen recordings in the description.

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift (hunk 3 above)
Comment (anandrajeswaran) [reply ios/FluentUI/Bottom Sheet/BottomSheetController.swift:796]: I don't know of any differences or have an preferences - but it feels like we should since otherwise it's 2 identical ways of doing the same thing? Can talk about outside the scope of this PR though.

Comment (rdeassis) [reply ios/FluentUI/Bottom Sheet/BottomSheetController.swift:796]: I think the undocumented pattern we're following is that we use computed properties when:
1. no parameters are needed
2. there is a return type
3. the implementation does not imply actions, but calculates a value to be returned

**Examples of computed properties:**
 - var leadingViewPadding: Int
 - var window: UIWindow?
 
 **Examples of funcs with no parameters:**
 - func updateForCurrentTheme()
 - func reload()
 - func setupDraggingWindow()

File: ios/FluentUI/Bottom Sheet/BottomSheetController.swift [hunk 4]
```diff
@@ -620,7 +648,7 @@ public class BottomSheetController: UIViewController {
             ? Constants.Spring.oscillatingDampingRatio
             : Constants.Spring.defaultDampingRatio
 
-        let springParams = UISpringTimingParameters(dampingRatio: damping, initialVelocity: CGVector(dx: 0.0, dy: springVelocity))
+        let springParams = UISpringTimingParameters(dampingRatio: damping, initialVelocity: CGVector(dx: springVelocity, dy: springVelocity))
```
Comment (rdeassis) [inline ios/FluentUI/Bottom Sheet/BottomSheetController.swift:651]: nit: break this into multiple lines for better readability?

**Review by rdeassis — APPROVED** [review]
